providers/dns/bluecat/bluecat.go

```diff
@@ -0,0 +1,407 @@
+// Package bluecat implements a DNS provider for solving the DNS-01 challenge
+// using a self-hosted Bluecat Address Manager.
+package bluecat
+
+import (
+	"bytes"
+	"strings"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+	"regexp"
+
+	"github.com/xenolf/lego/acme"
+)
+
+const bluecatUrlTemplate = "%s/Services/REST/v1"
+const TYPE_CONFIG = "Configuration"
+const TYPE_VIEW   = "View"
+const TYPE_TXT    = "TXTRecord"
+const TYPE_ZONE   = "Zone"
+
+type entityResponse struct {
+	Id         uint   `json:"id"`
+	Name       string `json:"name"`
+	Type       string `json:"type"`
+	Properties string `json:"properties"`
+}
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface that uses
+// Bluecat's Address Manager REST API to manage TXT records for a domain.
+type DNSProvider struct {
+	baseUrl      string
+	userName     string
+	password     string
+	configName   string
+	dnsView      string
+	token        string
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for Bluecat DNS.
+// Credentials must be passed in the environment variables: BLUECAT_SERVER_URL,
+// BLUECAT_USER_NAME and BLUECAT_PASSWORD. BLUECAT_SERVER_URL should have the
+// scheme, hostname, and port (if required) of the authoritative Bluecat BAM
+// server. The REST endpoint will be appended. In addition, the Configuration name
+// and external DNS View Name must be passed in BLUECAT_CONFIG_NAME and
+// BLUECAT_DNS_VIEW
+func NewDNSProvider() (*DNSProvider, error) {
+	server := os.Getenv("BLUECAT_SERVER_URL")
+	userName := os.Getenv("BLUECAT_USER_NAME")
+	password := os.Getenv("BLUECAT_PASSWORD")
+	configName := os.Getenv("BLUECAT_CONFIG_NAME")
+	dnsView := os.Getenv("BLUECAT_DNS_VIEW")
+	return NewDNSProviderCredentials(server, userName, password, configName, dnsView)
+}
+
+// NewDNSProviderCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for Bluecat DNS.
+func NewDNSProviderCredentials(server, userName, password, configName, dnsView string) (*DNSProvider, error) {
+	if server == "" || userName == "" || password == "" || configName == "" || dnsView == "" {
+		return nil, fmt.Errorf("Bluecat credentials missing")
+	}
+
+	return &DNSProvider{
+		baseUrl:      fmt.Sprintf(bluecatUrlTemplate, server),
+		userName:     userName,
+		password:     password,
+		configName:   configName,
+		dnsView:	      dnsView,
+	}, nil
+}
+
+//Send a REST request, using query parameters specified. The Authorization
+//header will be set if we have an active auth token
```

Comment:
I'm surprised `go fmt` didn't pick this up but generally godoc should have formatted with a space between the `//` and the comment text (like the rest of the code in this PR and repo)